### PR TITLE
feat: update network header based on the updated anonymousId

### PR DESCRIPTION
**Pull request by 1abhishekpandey**
## Description
This PR implements a feature to extracts `anonymousId` from event batches and dynamically updates the HTTP headers.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactor/optimization

## Implementation Details
- **EventUploader.swift**: Added anonymousId batch processing logic with regex-based extraction
  - New `lastBatchAnonymousId` property to track anonymous ID changes between batches
  - `updateAnonymousIdHeaderIfNeeded()` method to conditionally update headers
  - `extractAnonymousIdFromBatch()` method using NSRegularExpression for JSON parsing
  - Integration with existing upload pipeline in `uploadBatch()` method
- **HttpClient.swift**: Enhanced HTTP client to support dynamic anonymousId header updates
  - New `anonymousIdHeader` property to maintain current anonymous ID state
  - Modified `headers()` method to accept anonymousIdHeader parameter
  - Added `updateAnonymousIdHeader()` method for runtime header updates
  - Updated request preparation to use dynamic anonymousId instead of static analytics.anonymousId
- **Test Coverage**: Comprehensive unit tests for both components
  - `EventUploaderTests.swift`: 8 test cases covering various anonymousId extraction scenarios
  - `HttpClientTests.swift`: 6 test cases validating header management and configuration
  - Tests cover edge cases: malformed JSON, empty strings, special characters, complex payloads
- **Project Configuration**: Updated Xcode project file to include new test files and organize network tests

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?
1. **Unit Tests**: Run the test suite to verify anonymousId extraction and header management
   ```bash
   swift test --filter EventUploaderTests
   swift test --filter HttpClientTests
   ```
2. **Integration Testing**: Create events with different anonymousIds and verify batch processing
   - Initialize analytics with a specific anonymousId
   - Track multiple events and observe batch uploads
   - Verify that HTTP headers are updated when anonymousId changes between batches
3. **Edge Case Testing**: Test with malformed JSON, empty anonymousIds, and special characters
4. **Performance Testing**: Verify that regex operations don't introduce significant latency during batch processing

## Breaking Changes
None. This is a backward-compatible enhancement that extends existing functionality without modifying public APIs.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)
N/A - This is a backend/SDK enhancement with no UI changes.

## Additional Context
This implementation addresses the need for more efficient event batching by organizing events based on user anonymousId. The regex-based extraction approach was chosen for performance and reliability when parsing JSON batch payloads. The feature maintains backward compatibility while providing enhanced tracking capabilities for analytics processing.

Related to SDK optimization efforts and improves the reliability of user journey tracking across event batches. The implementation includes robust error handling and comprehensive test coverage to ensure stability in production environments.

## Summary by CodeRabbit

- **New Features**
  - Event uploads now include and propagate the latest AnonymousId per batch; headers are updated when the batch AnonymousId changes to ensure accurate tracking.
  - Batch processing now extracts AnonymousId from payloads before upload and caches it for subsequent requests.

- **Tests**
  - Added unit tests covering header construction, AnonymousId update flows and fallback behavior.
  - Added tests validating AnonymousId extraction across varied payload formats, ordering, whitespace, special characters, duplicates, and malformed inputs.

### CODE REVIEW

File: Sources/RudderStackAnalytics/EventQueue/EventUploader.swift [hunk 1]
```diff
@@ -50,6 +51,9 @@ final class EventUploader {
                         continue
                     }
                     
+                    // Update anonymousId header if needed
+                    self.updateAnonymousIdHeaderIfNeeded(batch)
+
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/EventQueue/EventUploader.swift:56]: It's guaranteed that batch will always contain the anonymousId.

File: Sources/RudderStackAnalytics/Network/HttpClient.swift [hunk 2]
```diff
@@ -96,17 +102,16 @@ enum HttpClientRequestType {
...
         if self == .events {
-            var specialHeaders = ["AnonymousId": analytics.anonymousId]
+            var specialHeaders = ["AnonymousId": anonymousIdHeader]
             if analytics.configuration.gzipEnabled { specialHeaders["Content-Encoding"] = "gzip" }
             specialHeaders.forEach { defaultHeaders[$0] = $1 }
         }
-        
+
         return defaultHeaders
     }
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Network/HttpClient.swift:116]: Ideally, that value should not be an empty one!

File: Sources/RudderStackAnalytics/EventQueue/EventUploader.swift [hunk 3]
```diff
@@ -144,4 +147,35 @@ extension EventUploader {
...
+            let anonymousIdRegex = try NSRegularExpression(pattern: "\"anonymousId\"\\s*:\\s*\"([^\"]+)\"")
+            let range = NSRange(location: 0, length: batch.utf16.count)
+
+            guard let match = anonymousIdRegex.firstMatch(in: batch, options: [], range: range) else {
+                return nil
+            }
+
+            guard let anonymousIdRange = Range(match.range(at: 1), in: batch) else {
+                return nil
+            }
```
Comment (SKannaniOS) [inline Sources/RudderStackAnalytics/EventQueue/EventUploader.swift]: Combine these two guard statements into one.

File: Sources/RudderStackAnalytics/Network/HttpClient.swift [hunk 4]
```diff
@@ -65,6 +67,10 @@ extension HttpClient: HttpClientRequests {
         
         return await HttpNetwork.perform(request: urlRequest).eventUploadResult
     }
+
+    func updateAnonymousIdHeader(_ anonymousIdHeader: String) {
+        self.anonymousIdHeader = anonymousIdHeader
+    }
```
Comment (SKannaniOS) [inline Sources/RudderStackAnalytics/Network/HttpClient.swift]: Move this method into the main class instead of keeping it in the extension related to `HttpClientRequests` protocol for better clarity.

File: Tests/RudderStackAnalyticsTests/EventQueue/EventUploaderTests.swift
Comment (SKannaniOS) [inline Tests/RudderStackAnalyticsTests/EventQueue/EventUploaderTests.swift:1]: Use Swift's `Testing` framework instead of `XCTest` framework.

File: Tests/RudderStackAnalyticsTests/Network/HttpClientTests.swift
Comment (SKannaniOS) [inline Tests/RudderStackAnalyticsTests/Network/HttpClientTests.swift:1]: Use Swift's `Testing` framework instead of `XCTest` framework.

**Review by SKannaniOS — CHANGES_REQUESTED** [review]

**Review by SKannaniOS — APPROVED** [review]
Nice.

**Review by vgupta98 — APPROVED** [review]
